### DISCUSSION TITLE
Intended behaviour for `externalDependencies`

### DISCUSSION THREAD

**Discussion by mischnic** [babel/babel#14727] [Q&A]
[I want to add support for `addExternalDependency` to Parcel](https://github.com/parcel-bundler/parcel/pull/7820), but found out that the meaning of these dependency strings isn't really specified.

https://github.com/babel/babel/pull/14065#issue-1084237130:
> `@babel/core` does not resolve external dependencies: they can be any string you want. This makes it possible, for example, to also store URLs or pointers to a virtual fs (as long as the caller knows how to handle them).

And the only tool I could find using this information was `babel-cli`, which does `path.resolve(process.cwd(), x)`:
https://github.com/babel/babel/blob/7d63d2f8337065a000d80da3282fba91f10fc6c2/packages/babel-cli/src/babel/watcher.ts#L74-L76

(Is anyone aware of any other tool reading `externalDependencies`?)

---

How should Parcel interpret these strings?
- For absolute paths, it's clear.
- For relative paths: what should they be relative to? To the babel config, to the JS file being processed, to some root directory (which I think babel-cli does)
- For anything else (though the line between "file that doesn't exist" and "not a filepath" isn't clear cut), I guess Parcel would ignore them.

This goes against the "any string you want" idea, but there has to be some agreement to be able to use plugins and runners interchangeably.

**Comment by JLHwung** [comment]
> For relative paths: what should they be relative to?

Good question! I think `process.cwd()` is not a bad default. Another choice is to respect `cwd` provided in Babel options. @nicolo-ribaudo Any idea?

**Reply by mischnic** [reply]
For Parcel, the cwd should be more or less irrelevant because running `parcel build src/index.js` vs `cd src; parcel build index.js` should both be equivalent. That's why I would prefer something that's based on the source file path (or config file path, ...)

**Reply by mischnic** [reply]
@nicolo-ribaudo Would be great to hear your thoughts 😄

**Reply by goatandsheep** [reply]
I'm also debugging a similar issue and have the same question.

**Reply by goatandsheep** [reply]
@JLHwung any way we can get expedited on this? I'm happy to share what I learn as open source docs but really struggling with this new feature

**Reply by liuxingbaoyu** [reply]
@goatandsheep Just a little curious, how is this problem bothering you? From my simple understanding, this seems to be an undefined behavior that may cause some temporary distress, but it doesn't seem to cause real blocking.

**Reply by goatandsheep** [reply]
Hi @liuxingbaoyu I'm not sure if it's this problem specifically, but my attempt to integrate `addExternalDependency` was unsuccessful at preventing caching of a non-js file that I attempted to add as an external dependency. When I tried `api.cache.never`, it caused an infinite loop, which was also wrong. I'm not sure if it's a path issue or not https://github.com/goatandsheep/react-native-dotenv/pull/288

**Reply by liuxingbaoyu** [reply]
@goatandsheep Hi! Have you tried converting all relative paths to absolute paths? This should avoid the problem.

**Reply by goatandsheep** [reply]
I will try this and get back to you

**Reply by goatandsheep** [reply]
I added path.join to make them absolute paths and noticed _something_ but the issue with the file caching remains. When I added path.join only to the addExternalDependency, it noticed the cache was not setup properly. When I added it to the cache.using the warning went away. So something is working with the addExternalDependency function. However the fact remains that when I update the file, the cache is not updating in my demo app
